documentation/rir.md

```diff
@@ -0,0 +1,133 @@
+  cat(as.character(i))  | ldfun "as.character" -> %0.0; ldvar "i" -> %0.1; call %0.0(%0.1) -> %0.2; ldfun "cat" -> %0.0; call %0.0(%0.2) -> %0.3
+}                       | > ...; branch BB3 BB2
+cat("]")
+```
+
+```r
+cat("[")                | ldfun "cat" -> %0.0; push "[" -> %0.1; call %0.0(%0.1) -> %0.2
+for (i in 1:3) {        | > [BB2] push 3; push 1; colon; forSeqSize; ...
+  cat(as.character(i))  | ldfun "as.character" -> %0.0; ldvar "i" -> %0.1; call %0.0(%0.1) -> %0.2; ldfun "cat" -> %0.0; call %0.0(%0.2) -> %0.3
+}                       | ...; branch BB3 BB2
+cat("]")
+```
+
+```r
+cat("[")                | ldfun "cat" -> %0.0; push "[" -> %0.1; call %0.0(%0.1) -> %0.2
+for (i in 1:3) {        | [BB2] push 3; push 1; colon; forSeqSize; ...
+  cat(as.character(i))  | > ldfun "as.character" -> %0.0; ldvar "i" -> %0.1; call %0.0(%0.1) -> %0.2; ldfun "cat" -> %0.0; call %0.0(%0.2) -> %0.3
+}                       | ...; branch BB3 BB2
+cat("]")
+```
+
+```r
+cat("[")                | ldfun "cat" -> %0.0; push "[" -> %0.1; call %0.0(%0.1) -> %0.2
+for (i in 1:3) {        | [BB2] push 3; push 1; colon; forSeqSize; ...
+  cat(as.character(i))  | ldfun "as.character" -> %0.0; ldvar "i" -> %0.1; call %0.0(%0.1) -> %0.2; ldfun "cat" -> %0.0; call %0.0(%0.2) -> %0.3
+}                       | > ...; branch BB3 BB2
+cat("]")
+```
+
+```r
+cat("[")                | ldfun "cat" -> %0.0; push "[" -> %0.1; call %0.0(%0.1) -> %0.2
+for (i in 1:3) {        | [BB2] push 3; push 1; colon; forSeqSize; ...
+  cat(as.character(i))  | ldfun "as.character" -> %0.0; ldvar "i" -> %0.1; call %0.0(%0.1) -> %0.2; ldfun "cat" -> %0.0; call %0.0(%0.2) -> %0.3
+}                       | ...; branch BB3 BB2
+> cat("]")              | > ldfun "cat" -> %0.0; push "[" -> %0.1; call %0.0(%0.1) -> %0.2
+```
+
+## Code Structure
+
+Here are a few important files and functions:
+
+- `Compiler.cpp`: Converts the parsed R AST into RIR (we don't parse the AST ourselves)
+  - `compileSpecialCall`: Converts calls to certain, "special" functions (e.g. `+`, `for`) into their own bytecodes. This way we can interpret them faster.
+- `interp.cpp` - `evalRirCode`: Contains the entire RIR interpreter
+
+## Comparison to GNU-R
+
+The default R interpreter (GNU-R) also has a bytecode. The main difference between this bytecode and RIR is that GNU-R has a few "fat" instructions, which are more complicated, while RIR has many more instructions, but they're simpler. For example, RIR has explicit instructions for creating environments, but GNU-R doesn't.
+
+## What about PIR?
+
+PIR is another representation of R, similar to RIR. However, PIR code isn't interpreted - instead, it's designed to be optimized. While interpreting, compiler converts RIR code into PIR, then performs some optimization passes, then converts the PIR back into RIR - *assuming the compiler works*, the new RIR code is semantically the same as the old, but faster.
```

Comment:
"similar to RIR" is maybe a bit misleading.

RIR is a stack-based bytecode. PIR is an SSA based IR.

Reply:
Maybe here the best is just to put a simple sentence and just link to pir documentation: in `documentation/`